feat(l2): implement /health for based components

FredPhilipy opened the pull request:
Implements health checks for `StateUpdater` and `BlockFetcher` components that were skipped in #4415.

- Add `CallMessage::Health` and health response structs for both components
- Modify `spawn()` methods to return `GenServerHandle` for admin server integration  
- Update admin server to include based components in `/health` endpoint
- Add integration test for health check serialization

**Health info includes:**
- ETH client connectivity status
- Component configuration (addresses, intervals)
- Current sequencer state

Closes #4508

LeanSerra commented on crates/l2/sequencer/admin_server.rs:
```diff
@@ -218,6 +228,28 @@ async fn health(
         );
     }
 
+    response.insert(
+        "state_updater".to_string(),
+        genserver_health(admin.state_updater, StateUpdaterCallMessage::Health, |msg| {
+            Some(match msg {
+                StateUpdaterOutMessage::Health(h) => h,
+                _ => return None,
+            })
+        })
+        .await,
+    );
+
+    response.insert(
+        "block_fetcher".to_string(),
+        genserver_health(admin.block_fetcher, BlockFetcherCallMessage::Health, |msg| {
+            Some(match msg {
+                BlockFetcherOutMessage::Health(h) => h,
+                _ => return None,
+            })
+        })
+        .await,
+    );
+
```
We should not insert into the response the health for based components if --based flag is not set. If we do we always get the message: `"block_fetcher": "Admin server does not have the genserver handle. Maybe it's not running?",`, which can be really confusing for operators that don't know that those components are only present  when using based sequencing and can lead to a false alarm being raised

LeanSerra commented on crates/l2/tests/test_based_health.rs:
Let's remove this file

LeanSerra commented on test_health_check.rs:
This shouldn't be on the root of the repo, let's remove it.

FredPhilipy commented:
> Hi! Thanks for the contribution, functionality looks great but left some comments. Also the CI is not passing

Thanks, made some fix from comment and the CI looks good now.

FredPhilipy commented:
Hi @LeanSerra is this good to go?

ElFantasma commented on crates/l2/sequencer/mod.rs:
```diff
@@ -157,8 +160,8 @@ pub async fn start_l2(
     }
-    if cfg.based.enabled {
-        let _ = StateUpdater::spawn(
+    let (state_updater, block_fetcher) = if cfg.based.enabled {
```
On current `main`, `StateUpdater::spawn()` already returns `GenServerHandle<StateUpdater>` and is already wired into the admin server. The `let (state_updater, block_fetcher) = if cfg.based.enabled { ... }` restructuring will conflict heavily — after rebase, only the `BlockFetcher` handle needs to be captured and threaded through to `start_api`.

ElFantasma commented on crates/l2/sequencer/admin_server.rs:
```diff
@@ -1,3 +1,11 @@
+use crate::based::{
```
On current `main`, `admin_server.rs` already imports `StateUpdater` and `StateUpdaterCallMessage`, and the `Admin` struct already has `state_updater: Option<GenServerHandle<StateUpdater>>`. Only the `BlockFetcher`-related imports and the block_fetcher health block in the `health()` handler are still needed after rebase.

ElFantasma commented on crates/l2/based/state_updater.rs:
```diff
@@ -1,15 +1,15 @@
-use std::{sync::Arc, time::Duration};
+use std::{collections::BTreeMap, sync::Arc, time::Duration};
```
All changes in this file are already on `main` (and the file has moved to `crates/l2/sequencer/state_updater.rs`). The current `main` version already has `CallMessage` (with `StopAt` variant), `handle_call`, `GenServerHandle` return from `spawn()`, and more. This entire file diff can be dropped after rebase.

FredPhilipy commented:
> This PR is heavily stale (Sep 2025) and roughly half of it has already been merged into `main`. Specifically, `StateUpdater` on current `main` already has `CallMessage`, `handle_call`, returns `GenServerHandle` from `spawn()`, and is wired into the admin server — so all the `state_updater.rs` and most of the `sequencer/mod.rs` changes are already superseded.
> 
> The remaining value is the **BlockFetcher health check** (adding `CallMessage::Health`, `BlockFetcherHealth` struct, `handle_call`, returning `GenServerHandle` from `spawn()`, and wiring it into the `/health` endpoint). This part is still needed on current `main`. A rebase is required — the diff will shrink significantly and there will be conflicts in `sequencer/mod.rs` and `admin_server.rs` due to restructuring.
> 
> **crates/l2/sequencer/mod.rs** — On current `main`, `BlockFetcher::spawn()` returns `Result<(), BlockFetcherError>` and the result is discarded with `let _ =`. The PR changes it to return `GenServerHandle`, but the surrounding code structure in `start_l2` has changed significantly (e.g., `state_updater` is already handled differently). This file will need careful conflict resolution.
> 
> **crates/l2/based/block_fetcher.rs** — The `HashMap` import in the original diff adds `BTreeMap` alongside `HashMap`, but on current `main` the `HashMap` import was removed (the file only uses `std::cmp::min`). After rebase, only `BTreeMap` needs to be added.

Thanks for the detailed review @ElFantasma! You're right, the PR was heavily stale.

I've rebased onto current main and cleaned up the diff:

Dropped all StateUpdater structural changes that were already merged (spawn returning handle, CallMessage, file relocation to sequencer/)
Added Health variant to StateUpdater's existing CallMessage enum and the corresponding StateUpdaterHealth struct
Kept BlockFetcher health check as the main new addition (CallMessage, Health struct, handle_call, spawn returning handle)
Updated admin_server.rs to include both components in /health — only the new health blocks, built on top of the existing genserver_health helper
The diff went from ~162 to ~130 lines across 4 files
Should be much cleaner now. Let me know if anything else needs adjustment!

FredPhilipy commented on crates/l2/sequencer/state_updater.rs:
```diff
@@ -297,6 +310,19 @@ impl StateUpdater {
+            sequencer_registry_address: self.sequencer_registry_address,
+            sequencer_address: self.sequencer_address,
+            check_interval_ms: self.check_interval_ms,
+            sequencer_state: format!("{:?}", self.sequencer_state.status()),
```
Same as above — status() is sync, no .await needed.
